File: app/src/main/java/com/example/weatherapp/FcstResult.java
package com.example.weatherapp;

import java.util.Collections;
import java.util.List;

public class FcstResult implements APIResult<WeatherData> {
    Response response;

    @Override
    public WeatherData toAppData() throws Exception {
        if (!response.header.resultCode.equals("00")) throw new Exception();

        WeatherData weatherData = new WeatherData();

        List<Item> listItem = response.body.items.item;

        Collections.sort(listItem);

        String fcstTime = listItem.get(0).fcstTime;
        WeatherData.WeatherHour weatherHour = new WeatherData.WeatherHour();
        weatherHour.fcstDate = listItem.get(0).fcstDate;
        weatherHour.fcstTime = listItem.get(0).fcstTime.substring(0, 2) + ":" + listItem.get(0).fcstTime.substring(2);

        for (Item item : listItem) {
            if (!fcstTime.equals(item.fcstTime)) {
                weatherData.addWeatherHour(weatherHour);
                fcstTime = item.fcstTime;

                weatherHour = new WeatherData.WeatherHour();
                weatherHour.fcstDate = item.fcstDate;
                weatherHour.fcstTime = item.fcstTime.substring(0, 2) + ":" + item.fcstTime.substring(2);
            }

            if (item.category.equals(Category.TMN.getValue())) {
                int dayDiff = Integer.parseInt(item.fcstDate) - Integer.parseInt(item.baseDate);
//                weatherData.tmn[dayDiff] = Double.parseDouble(item.fcstValue);
            }
            else if (item.category.equals(Category.TMX.getValue())) {
                int dayDiff = Integer.parseInt(item.fcstDate) - Integer.parseInt(item.baseDate);
//                weatherData.tmx[dayDiff] = Double.parseDouble(item.fcstValue);
            }
            else {
                weatherHour.setFcstValue(item.category, item.fcstValue);
            }
        }

        weatherData.addWeatherHour(weatherHour);
        return weatherData;
    }

    static class Response {
        Header header;
        Body body;
    }

    static class Header {
        String resultCode;
        String resultMsg;
    }

    static class Body {
        String dataType;
        int numOfRows;
        int pageNo;
        int totalCount;
        Items items;
    }

    static class Items {
        List<Item> item;
    }

    static class Item implements Comparable<Item> {
        String baseDate;
        String baseTime;
        String category;
        String fcstDate;
        String fcstTime;
        String fcstValue;
        int nx;
        int ny;

        @Override
        public int compareTo(Item o) {
            String o1DateTime = fcstDate + fcstTime;
            String o2DateTime = o.fcstDate + o.fcstTime;
            return o1DateTime.compareTo(o2DateTime);
        }
    }
}

File: app/src/main/java/com/example/weatherapp/data/source/remote/fcst/FcstResult.kt
package com.example.weatherapp.data.source.remote.fcst

class FcstResult {
    var response: Response? = null

    class Response {
        var header: Header? = null
        var body: Body? = null
    }

    class Header {
        var resultCode: String? = null
        var resultMsg: String? = null
    }

    class Body {
        var dataType: String? = null
        var numOfRows = 0
        var pageNo = 0
        var totalCount = 0
        var items: Items? = null
    }

    class Items {
        var item: MutableList<Item>? = null
    }

    class Item : Comparable<Item> {
        var baseDate: String? = null
        var baseTime: String? = null
        var category: String? = null
        var fcstDate: String? = null
        var fcstTime: String? = null
        var fcstValue: String? = null
        var nx = 0
        var ny = 0
        override fun compareTo(other: Item): Int {
            val o1DateTime = fcstDate + fcstTime
            val o2DateTime = other.fcstDate + other.fcstTime
            return o1DateTime.compareTo(o2DateTime)
        }
    }
}